extractors/cds/tools/src/cds/compiler/graph.ts

```diff
@@ -0,0 +1,476 @@
+): CompilationAttempt {
+  const attemptId = `${task.id}_${Date.now()}_${Math.random().toString(36).substr(2, 5)}`;
+  const startTime = new Date();
+
+  const attempt: CompilationAttempt = {
+    id: attemptId,
+    cdsCommand,
+    cacheDir,
+    timestamp: startTime,
+    result: {
+      success: false,
+      timestamp: startTime,
+    },
+  };
+
+  try {
+    // For now, we'll use the first source file for compilation
+    // In a more sophisticated implementation, we might handle project-level compilation differently
+    const primarySourceFile = task.sourceFiles[0];
+
+    const compilationResult = compileCdsToJson(
+      primarySourceFile,
+      dependencyGraph.sourceRootDir,
+      cdsCommand,
+      cacheDir,
+      // Convert CDS projects to BasicCdsProject format expected by compileCdsToJson
+      new Map(
+        Array.from(dependencyGraph.projects.entries()).map(([key, value]) => [
+          key,
+          {
+            cdsFiles: value.cdsFiles,
+            cdsFilesToCompile: value.cdsFilesToCompile,
+            expectedOutputFiles: value.expectedOutputFiles,
+            projectDir: value.projectDir,
+            dependencies: value.dependencies,
+            imports: value.imports,
+            packageJson: value.packageJson,
+            compilationConfig: value.compilationConfig,
+          },
+        ]),
+      ),
+      task.projectDir,
+    );
+
+    const endTime = new Date();
+    attempt.result = {
+      ...compilationResult,
+      timestamp: endTime,
+      durationMs: endTime.getTime() - startTime.getTime(),
+      commandUsed: cdsCommand,
+      cacheDir,
+    };
+
+    if (compilationResult.success && compilationResult.outputPath) {
+      dependencyGraph.statusSummary.jsonFilesGenerated++;
+    }
+  } catch (error) {
+    const endTime = new Date();
+    attempt.error = {
+      message: String(error),
+      stack: error instanceof Error ? error.stack : undefined,
+    };
+    attempt.result.timestamp = endTime;
+    attempt.result.durationMs = endTime.getTime() - startTime.getTime();
+  }
+
+  task.attempts.push(attempt);
+  return attempt;
+}
+
+/**
+ * Create a compilation task for a project or individual file
+ */
+function createCompilationTask(
+  type: 'file' | 'project',
+  sourceFiles: string[],
+  expectedOutputFiles: string[],
+  projectDir: string,
+  useProjectLevelCompilation: boolean,
+  priority: number = 0,
+): CompilationTask {
+  return {
+    id: `${type}_${projectDir}_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`,
+    type,
+    status: 'pending',
+    sourceFiles,
+    expectedOutputFiles,
+    projectDir,
+    attempts: [],
+    useProjectLevelCompilation,
+    priority,
+    dependencies: [],
+  };
+}
+
+/**
+ * Create compilation configuration with retry alternatives
+ */
+function createCompilationConfig(
+  primaryCommand: string,
+  primaryCacheDir: string | undefined,
+  useProjectLevel: boolean,
+  alternatives: AlternativeCdsCommand[] = [],
```

Comment:
Just noting that this is unused presently. Do we need to retain this for any future purpose?